src/System.Buffers/src/System/Buffers/ArrayPoolBucket.cs

```diff
@@ -0,0 +1,104 @@
+        private T[][] _data;
+        private SpinLock _lock;
+
+        /// <summary>
+        /// Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.
+        /// </summary>
+        internal ArrayPoolBucket(int bufferLength, int numberOfBuffers)
+        {
+            _index = 0;
+            _lock = new SpinLock();
+            _data = new T[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new T[bufferLength];
+            }
+        }
+
+        /// <summary>
+        /// Returns an array from the Bucket sized according to the Bucket size.
+        /// If the Bucket is empty, null is returned.
+        /// </summary>
+        /// <returns>Returns a valid buffer when the bucket has free buffers; otherwise, returns null</returns>
+        internal T[] Rent()
+        {
+            T[] buffer;
+
+            // Use a SpinLock since it is super lightweight
+            // and our lock is very short lived
+            bool taken = false;
+            _lock.Enter(ref taken);
+            Debug.Assert(taken);
+
+            // Check if all of our buffers have been rented
+            if (_index >= _data.Length)
+            {
+                // Exit the lock as soon as possible
+                _lock.Exit(false);
+                return null;
+            }
+            else
+            {
+                // We still have buffers so grab one and return it
+                buffer = _data[_index];
+#if DEBUG
+                // Only set to NULL in debug mode to track usage, otherwise 
+                // we can skip this instruction to get a minimal speed increase
+                _data[_index] = null;
+#endif
+                _index++;
+
+                // Exit the lock as soon as possible
+                _lock.Exit(false);
+            }
+
+            return buffer;
+        }
+
+        /// <summary>
+        /// Attempts to return a Buffer to the bucket. This can fail
+        /// if the buffer being returned was allocated and we don't have
+        /// room for it in the bucket.
+        /// </summary>
+        /// <returns>Returns true when the buffer is successfull returned; otherwise, returns false</returns>
+        internal bool Return(T[] buffer)
+        {
+            // Use a SpinLock since it is super lightweight
+            // and our lock is very short lived
+            bool taken = false;
+            _lock.Enter(ref taken);
+            Debug.Assert(taken);
+
+            // If we have space to put the buffer back, do it. If we don't
+            // then there was a buffer alloc'd that was returned instead so
+            // we can just drop this buffer
+            bool returned = false;
+            if (_index != 0)
+            {
+                _index--;
+                _data[_index] = buffer;
+                returned = true;
+            }
+
+            _lock.Exit(false);
```

Comment:
Is there a standard way to harden against that scenario? I haven't done that before